Print fixes

vadz opened the pull request:
Restore compatibility with the existing code overriding `GetPageInfo()` by ignoring values returned by it (sic).

Also a couple of minor wxMSW-specific printing-related bug fixes/improvements.

vadz committed:
Check for HDC validity at the start of StartDoc()

No real changes, just check the precondition before doing anything else
for consistency with the other functions and because it doesn't make
sense to try doing anything if we don't have a valid HDC anyhow.

vadz committed:
Fix printing selected pages when overriding GetPageInfo()

Simply ignore the "from" and "to" values returned from this legacy
function as wxMSW always did it and nothing else makes much sense, as we
don't want to override the page range(s) entered by the user in the
print dialog.

Update the documentation to clarify this unexpected behaviour and also
provide a better explanation of what the new GetPagesInfo() should, and
should not, do.

Closes #25254 (again).

vadz committed:
Avoid giving an error after cancelling file dialog when printing

Cancelling the dialog shown to select the file to print to, when
printing to a file, shouldn't result in an error message, so fix this by
checking last error after calling Win32 StartDoc() which does it.

As there is no clean way to pass this error from wxPrinterDC::StartDoc()
back to wxWindowsPrinter, which calls it indirectly via wxPrintout, just
pass it via a global variable -- this is ugly, but should work fine as
we can't be printing more than one thing at the same time.

vadz committed:
Avoid bogus SetFocus() debug errors while showing print dialog

Getting WM_ACTIVATE just before the print dialog dismissal but while it
still existed resulted in spurious errors about not being able to
restore the last focus in the previously active window because it didn't
recover its active status yet, at this moment.

Use an ugly hack to ignore activation events happening in this case, as
there doesn't seem to be anything else that we can do to avoid getting
them.

Closes #13089.

taler21 commented:
Thanks, but "Print all pages" doesn't work correctly.

The cause is that `wxWindowsPrintDialog::ConvertFromNative( wxPrintDialogData &data )` sets a `wxPrintPageRange`

https://github.com/wxWidgets/wxWidgets/blob/12b29979d3ba9a5a08b2b3e44ad775ebe3ea18af/src/msw/printdlg.cpp#L1013

and `wxWindowsPrinter::Print()` uses it,

https://github.com/wxWidgets/wxWidgets/blob/12b29979d3ba9a5a08b2b3e44ad775ebe3ea18af/src/msw/printwin.cpp#L152-L156

both regardless of whether the user actually selected "Print pages" (`PD_PAGENUMS`) in the print dialog.

So I think there are two possible solutions:

1. `ConvertFromNative()` calls `wxPrintDialogData::SetPageRanges()` passing an empty `wxPrintPageRange` vector if `PD_PAGENUMS` was not selected by the user in the print dialog.
2. `wxWindowsPrinter::Print()` only initializes page ranges with the value from the dialog, if "Print pages" was selected by the user.
    ```cpp
    std::vector<wxPrintPageRange> pageRanges;
    if ( (!m_printDialogData.GetAllPages()) && (!m_printDialogData.GetSelection()) && (!m_printDialogData.GetCurrentPage()) )
        pageRanges = m_printDialogData.GetPageRanges();
    ```

vadz committed:
Update comments related to page ranges in wxWindowsPrintDialog

Remove outdated comment saying that only one page range is supported
and explain that we always create a page range for compatibility.

No real changes.

vadz committed:
Fix regression with printing all pages in wxMSW

After the recent changes, only the first page was printed even when
"All" was selected in the print dialog because we used the page range
from the dialog by default.

Only use it if this is what was actually selected in the dialog.

Co-authored-by: taler21 <99262969+taler21@users.noreply.github.com>

vadz committed:
Add helper wxPrintDialogData::GetSpecifiedPages()

This is more convenient than testing that we're not printing all pages
or printing selection or printing the current page.

vadz commented:
> Thanks, but "Print all pages" doesn't work correctly.

Oops, sorry for not noticing that this was broken now :-(

> So I think there are two possible solutions:

I think the solution (2) is clearly better, we should still copy the page ranges specified by user (if only to restore them the next time the dialog is used, although the sample doesn't do this correctly and reuses only `wxPrintData`, but not `wxPrintDialogData`, which contains this information, for subsequent printings), but we shouldn't use them if this is not the active choice. I've implemented this now and pushed this change, thanks! (I just couldn't avoid the urge to add a helper function because the test above looks really ugly).

<sub>The rest is just some musings about problems not directly related to this issue and can be skipped, but I'd like to write this down at least somewhere.</sub>

-----

While looking at this code, I got confused once again by thinking that it would be correct to call `GetPagesInfo()` *before* showing the dialog because I wrongly thought that the values returned from it should be used to initialize the dialog. And this is not the first time I was confused by this, and it's not just me because I had applied a patch from @troels-k back in 41514cc474 (Set up pages range in the wxMSW print dialog correctly., 2011-01-03), see #12819, only to revert it very quickly afterwards in ff4713940d (Revert "Set up pages range in the wxMSW print dialog correctly.", 2011-01-29)., see #12911. And, to be complete, except for less than one month between these commits, this had been like this even since d2b354f9c9 (Printing improvements: GetPageInfo() gets called after the DC has been set and after OnPreparePrinting() has been called [...], 2003-07-23) and `OnPreparePrinting()` was moved after `GetPageInfo()` even before it, in 1044a386f0 (Misc. 16-bit compilation fixes, 1999-09-05).

And this is correct because this function is supposed to be able to use the printing DC to paginate the document correctly, but this DC can't be created before showing the dialog. So the values returned by it can't be used to initialize the dialog, and only the values from `wxPrintDialogData` can be. And this is, again, done wrongly in the sample: it should preserve them between calls. I've changed this now and updated the documentation to make this more clear.

vadz committed:
Replace wxPrintDialogData booleans with a single enum

It was unnatural to have 3 booleans encoding 4 different states, as this
made a lot of impossible states representable.

Change this by using a single 4-state enum instead.

The price to pay is that SetAllPages(false) etc now behave somewhat
strangely: they don't do anything when not printing all pages (same as
before) and set the print target to "only specified pages" otherwise.
This is almost but not quite the same as the old behaviour because
calling

        SetAllPages(true);
        SetSelection(true);
        SetAllPages(false);

resulted in GetSelection() being true before but GetSpecifiedPages()
being true now. But hopefully not many people did anything like this and
at least this makes the state after the first 2 calls well-defined,
unlike before.

Also make printing all pages default everywhere and not just under Mac
because nothing else seems to make sense.

Update the documentation and, in particular, add the documentation of
SetAllPages() which somehow was never documented before.
